### PR TITLE
Update sd-dhcp-client.c to support multiple DHCP servers 

### PR DESCRIPTION
**Pull request by alexandref75**
Reject NAKs from servers that we did not send a request to. That allows systemd-networkd to operate in networks that have multiple DHCP servers.

### CODE REVIEW

File: src/libsystemd-network/sd-dhcp-client.c [hunk 1]
```diff
@@ -1591,9 +1591,15 @@ static int client_parse_message(
         case DHCP_STATE_REQUESTING:
         case DHCP_STATE_RENEWING:
         case DHCP_STATE_REBINDING:
-                if (r == DHCP_NAK)
+                if (r == DHCP_NAK) {
+                        if ((client->lease != NULL) &&
+                            (client->lease->server_address != lease->server_address))
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-client.c]: Please drop redundant `()`. Also, we allow to implicit cast pointers to boolean. So, 
```c
if (client->lease && client->lease->server_address == lease->server_address)
        return log_dhcp_client_errno(...);
```

File: src/libsystemd-network/sd-dhcp-client.c [hunk 2]
```diff
@@ -1591,9 +1591,15 @@ static int client_parse_message(
...
+                                    return log_dhcp_client_errno(client, SYNTHETIC_ERRNO(ENOMSG),
+                                                                 "NAK: %s from other server",
```
Comment (yuwata) [inline src/libsystemd-network/sd-dhcp-client.c]: Please explicitly mention that the message will be ignored.
```c
"NAK: %s from other server, ignoring."
```

**Review by yuwata — CHANGES_REQUESTED** [review]
The basic direction looks good to me. Several coding style requests.

File: src/libsystemd-network/sd-dhcp-client.c (hunk 2 above)
Comment (alexandref75) [reply src/libsystemd-network/sd-dhcp-client.c]: Thanks, changed.

File: src/libsystemd-network/sd-dhcp-client.c (hunk 1 above)
Comment (alexandref75) [reply src/libsystemd-network/sd-dhcp-client.c]: Thanks, changed.

**Review by yuwata — APPROVED** [review]
LGTM. Thank you!

Next time, please squash commits and force-push revised commit(s).
This PR is small enough and we can squash them on merge.

**Comment by alexandref75** [comment]
Sorry for not catching it, but when I modified if statement the comparison was inverted. We want to reject all NAKs that do not belong to server we sent.

**Comment by yuwata** [comment]
> Sorry for not catching it, but when I modified if statement the comparison was inverted. We want to reject all NAKs that do not belong to server we sent.

Ouch. Sorry, that's my mistake. Thank you for catching that.

**Comment by mbiebl** [comment]
@yuwata this was requested to be backported to v252 downstream at [1]

The change looks small enough, wdyt?

[1] https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1052365#15

**Comment by yuwata** [comment]
> @yuwata this was requested to be backported to v252 downstream at [1]
> 
> The change looks small enough, wdyt?
> 
> [1] https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1052365#15

Sounds good. But, IIRC we do not actively release v252 point releases anymore except for critical issues (@bluca Is this correct?). So, if a downstream user requests to backport the commit, please do that in downstream side, or at least please open a pull request to stable branch.

**Comment by bluca** [comment]
We still do full releases for 252, but somebody needs to backport it, test it and send a PR

**Pull request merged**